app/graphql/concerns/fancy_mutation.rb

```diff
@@ -0,0 +1,196 @@
+
+    # The return-driven approach of #ready? and #authorized? is garbage, so we override it and allow
+    # using the same error system as in #resolve. To achieve this, we prepend a module wrapping the
+    # mutation's #ready? and #authorized? methods, detecting the state of the error object after
+    # execution and changing the result if needed.
+    def ready?(*)
+      ready, result = super
+
+      return [false, { errors: errors }] if errors.present?
+
+      [ready, result]
+    end
+
+    def authorized?(*)
+      ready, result = super
+
+      return [false, { errors: errors }] if errors.present?
+
+      [ready, result]
+    end
+  end
+
+  class_methods do
+    # Sets the result type
+    # @param type [Class] The GraphQL type to use for the result
+    def result(type, **kwargs)
+      field :result, type, **kwargs
+    end
+
+    # Sets up an input type and field for the mutation
+    # @param block [Proc] A block defining the input type fields
+    def input(&block)
+      input_type.instance_eval(&block)
+    end
+
+    # Adds types to the union of error types for the mutation
+    # @param types [Array<Class>] The error types this can return
+    def errors(*types)
+      errors_union.possible_types(*types)
+    end
+
+    # Adds types to the union of warning types for the mutation
+    # @param types [Array<Class>] The warning types this can return
+    def warnings(*types)
+      warnings_union.possible_types(*types)
+    end
+
+    ### private_class_methods ###
+
+    # These methods are kinda strange. The first time they're called, they create a new type, hook
+    # it up to the mutation, and return it. When they get called again, they return that same type,
+    # allowing us to modify them repeatedly from different locations. This allows utilities to add
+    # types automatically such as an ignore_warnings argument, or a common error type.
+    def errors_union
+      @errors_union ||= begin
+        union_name = "#{graphql_name}ErrorsUnion"
+        errors_union = Class.new(Types::Union::Base) do
+          graphql_name union_name
+
+          def self.resolve_type(object, _context)
+            object.key?(:__type) ? object[:__type] : super
+          end
+        end
+        field :errors, [errors_union], null: true
+
+        errors_union
+      end
+    end
+
+    def warnings_union
+      @warnings_union ||= begin
+        union_name = "#{graphql_name}WarningsUnion"
+        warnings_union = Class.new(Types::Union::Base) do
+          graphql_name union_name
+
+          def self.resolve_type(object, _context)
+            object.key?(:__type) ? object[:__type] : super
+          end
+        end
+        field :warnings, [warnings_union], null: true
+
+        # Add the ignore_warnings argument and error type
+        input { argument :ignore_warnings, GraphQL::Types::Boolean, required: false }
+        errors Types::Errors::WarningsPresent
+
+        warnings_union
+      end
+    end
+
+    def input_type
```

Comment:
This makes the input type reopenable, so we can call it multiple times to add more arguments